presto-main-base/src/main/java/com/facebook/presto/sql/MaterializedViewUtils.java

```diff
@@ -327,4 +340,142 @@ public boolean validate(Identifier baseTableColumn, Map<Expression, Identifier>
+                conjuncts.add(domainExpr);
+            }
+
+            if (!conjuncts.isEmpty()) {
+                Expression conjunction;
+                if (conjuncts.size() == 1) {
+                    conjunction = conjuncts.get(0);
+                }
+                else {
+                    // Chain multiple conjuncts with AND: (A AND B AND C)
+                    conjunction = conjuncts.stream()
+                            .reduce((left, right) -> new LogicalBinaryExpression(
+                                    LogicalBinaryExpression.Operator.AND, left, right))
+                            .get();
+                }
+                disjuncts.add(conjunction);
+            }
+        }
+
+        if (disjuncts.isEmpty()) {
+            throw new IllegalStateException("No predicates generated for missing partitions");
+        }
+
+        if (disjuncts.size() == 1) {
+            return disjuncts.get(0);
+        }
+        else {
+            return disjuncts.stream()
+                    .reduce((left, right) -> new LogicalBinaryExpression(
+                            LogicalBinaryExpression.Operator.OR, left, right))
+                    .get();
+        }
+    }
+
+    public static Expression domainToExpression(Expression columnRef, Domain domain)
+    {
+        if (domain.isNullAllowed() && domain.getValues().isNone()) {
+            return new IsNullPredicate(columnRef);
+        }
+
+        ValueSet values = domain.getValues();
+
+        List<Expression> valueExprs = new java.util.ArrayList<>();
+        for (com.facebook.presto.common.predicate.Range range : values.getRanges().getOrderedRanges()) {
+            if (range.isSingleValue()) {
+                valueExprs.add(toExpressionLiteral(range.getSingleValue(), domain.getType()));
+            }
+            else {
+                throw new UnsupportedOperationException(
+                        "Range predicates not yet supported for auto-refresh. Found range: " + range);
+            }
+        }
+
+        if (valueExprs.isEmpty()) {
+            throw new IllegalStateException("No values found in domain");
+        }
+
+        if (valueExprs.size() == 1) {
+            return new ComparisonExpression(
+                    ComparisonExpression.Operator.EQUAL,
+                    columnRef,
+                    valueExprs.get(0));
+        }
+        else {
+            return new InPredicate(columnRef, new InListExpression(valueExprs));
+        }
+    }
+
+    public static Expression toExpressionLiteral(Object value, Type type)
+    {
+        if (value == null) {
+            return new NullLiteral();
+        }
+
+        if (type instanceof VarcharType) {
+            return new StringLiteral(((io.airlift.slice.Slice) value).toStringUtf8());
+        }
+
+        if (type instanceof BigintType || type instanceof IntegerType) {
+            return new LongLiteral(value.toString());
+        }
+
+        if (type instanceof DateType) {
```

Comment:
I'm not sure if this is right.  At the very least might be worthwhile to have an assertion on what we expect the `value` type to be here, as I don't know it will work if it's represented as a whole number (as it is in `LiteralEncoder`).